### PR TITLE
Reorg TLV handling and add test sequence for SMP TLVs

### LINKED ISSUE #5: handleDataMessage() ignores data if a TLV is present
The OTR spec describes a Data Message saying: "...a human-readable message..., optionally followed by... zero or more TLV (type/length/value) records..."[[1](https://otr.cypherpunks.ca/Protocol-v3-4.0.0.html)]  but `handleDataMessage()` always returns a `null` "human readable message" whenever a TLV is present:

``` java
                if (tlvs != null && tlvs.size() > 0) {
                    for (TLV tlv : tlvs) {
                        switch (tlv.getType()) {
                            case TLV.DISCONNECTED:
                                this.setSessionStatus(SessionStatus.FINISHED);
                                return null;
                            default:
                                if (otrSm.doProcessTlv(tlv))
                                    return null;
                        }
                    }
                }
                return decryptedMsgContent;
```

### LINKED ISSUE #11: handleDataMessage() ignores data if a TLV is present
The OTR spec describes a Data Message saying: "...a human-readable message..., optionally followed by... zero or more TLV (type/length/value) records..."[[1](https://otr.cypherpunks.ca/Protocol-v3-4.0.0.html)]  but `handleDataMessage()` always returns a `null` "human readable message" whenever a TLV is present:

```java
                if (tlvs != null && tlvs.size() > 0) {
                    for (TLV tlv : tlvs) {
                        switch (tlv.getType()) {
                            case TLV.DISCONNECTED:
                                this.setSessionStatus(SessionStatus.FINISHED);
                                return null;
                            default:
                                if (otrSm.doProcessTlv(tlv))
                                    return null;
                        }
                    }
                }
                return decryptedMsgContent;
```

### PR DESCRIPTION
**Pull request by eighthave**
This pull request is a Work In Progress, but I wanted to post it because I'm a bit stumped why the SMP test I wrote does not complete.  So I'm hoping someone can spot the problem.  Then I can fix it, rebase, and then this request will be ready for merging.

I wanted to write the SMP tests so we can reorganize the TLV handling and have something to make sure we didn't break things.  I already included one commit that reorgs the TLV handling and fixes #11 also.

I also purged some dead code, and added some javadoc comments on the public `interface`.

### CODE REVIEW

**Comment by cobratbq** [comment]
hi @eighthave I'll have a look at this this weekend. Then I'll have some more time to actually dive into the changes.

**Comment by cobratbq** [comment]
@eighthave, not cool man! You are generating a new keypair on every call to getLocalKeyPair()! (See DummyOtrEngineHostImpl)

Try something like this, that should fix your problem :+1: 

``` java
        private HashMap<SessionID, KeyPair> keypairs = new HashMap<SessionID, KeyPair>();

        public KeyPair getLocalKeyPair(SessionID paramSessionID) {
            KeyPair keypair = this.keypairs.get(paramSessionID);
            if (keypair == null) {
                KeyPairGenerator kg;
                try {
                    kg = KeyPairGenerator.getInstance("DSA");
                    keypair = kg.genKeyPair();
                    this.keypairs.put(paramSessionID, keypair);
                } catch (NoSuchAlgorithmException ex) {
                    logger.log(Level.SEVERE, null, ex);
                }
            }
            return keypair;
        }
```

On the positive side: your test succeeded immediately after having fixed that mistake :-D

EDIT: Okay, technically it wasn't your code that keeps generating new keypairs. Still, it was a really annoying problem :-P

**Comment by eighthave** [comment]
Ok, this is now complete, thanks to @cobratbq for the final tip that got the whole thing working.  I also added some more SMP tests that throw lots of unicode at the process.  The unicode chars seem to work fine, but it seems that special chars like `\n \t \r \0` in the question cause an exception, so I filed a bug: #30.

File: pom.xml [hunk 1]
```diff
@@ -104,6 +104,9 @@
                     <systemPropertyVariables>
                         <java.util.logging.SimpleFormatter.format>%1$tH:%1$tM:%1$tS
                             %4$-6s %2$s(): %3$s "%5$s%6$s"%n</java.util.logging.SimpleFormatter.format>
+                  <java.util.logging.config.file>
+                    ${basedir}/logging.properties
+                  </java.util.logging.config.file>
```
Comment (cobratbq) [inline pom.xml]: Fix indentation

File: src/main/java/net/java/otr4j/session/Session.java [hunk 2]
```diff
@@ -648,12 +648,29 @@ private String handleDataMessage(DataMessage data) throws OtrException {
                 if (tlvs != null && tlvs.size() > 0) {
                     for (TLV tlv : tlvs) {
                         switch (tlv.getType()) {
-                            case TLV.DISCONNECTED:
+                            // TVL.PADDING // TLV0
```
Comment (cobratbq) [inline src/main/java/net/java/otr4j/session/Session.java]: I think we should make this case explicit. So TLV.PADDING: break; That way, if ever we decide to have a "default" case for unsupported types, we can be sure that we _only_ catch the unsupported types.

**Comment by cobratbq** [comment]
I haven't checked the various processTlv\* methods yet. I may be able to do that tomorrow. Or maybe languitar beats me to it, which is fine too of course.

File: src/main/java/net/java/otr4j/session/Session.java (hunk 2 above)
Comment (eighthave) [inline src/main/java/net/java/otr4j/session/Session.java]: makes sense, fixed.  I also added a `default:` to throw a warning message to aid debugging.

File: pom.xml (hunk 1 above)
Comment (eighthave) [inline pom.xml]: fixed

**Comment by eighthave** [comment]
the various `processTlv`\* methods are just the unmodified code from before, but instead of being in an if-then-else ladder, they are individual methods.

**Comment by cobratbq** [comment]
Indeed, processTlv\* are just split into separate methods. I checked the rest, including a quick test use of the otr4j library and everything's in order AFAICT. Merging ...
